### PR TITLE
[Bugfix]handle errors returned by storage layer in scan operator

### LINKED ISSUE #6030: The value in MySQL exceeds the range of Sr, and the result is empty
### Steps to reproduce the behavior (Required)
1. mysql environment Server version: 8.0.29 MySQL
create database db_mysql_test;
use db_mysql_test;
create table test_tbl_char(
type_char char(9)
);
insert into test_tbl_char values('abcdefghi');

create table test_tbl_varchar(
type_varchar varchar(9)
);
insert into test_tbl_varchar values('abcdefghi');

create table test_tbl_int(
type_int int
);
insert into test_tbl_int values('2147483647');

2. sr environment
create external resource jdbc_ext_tbl_resource_outofrange
                properties (
                    "type"="jdbc",
                    "user"="root",
                    "password"="*******",
                    "jdbc_uri"="jdbc:mysql://xxx.xx.xx.xx:3306/db_mysql_test",
                    "driver_url"="https://repo1.maven.org/maven2/mysql/mysql-connector-java/8.0.29/mysql-connector-java-8.0.29.jar",
                    "driver_class"="com.mysql.jdbc.Driver"
                );

create external table jdbc_ext_table_mysql_char (
                    type_char char(8)
                    ) ENGINE=jdbc PROPERTIES (
                        "resource"="jdbc_ext_tbl_resource_outofrange",
                        "table"="test_tbl_char"
                    );

create external table jdbc_ext_table_mysql_varchar (
                    type_varchar varchar(8)
                    ) ENGINE=jdbc PROPERTIES (
                        "resource"="jdbc_ext_tbl_resource_outofrange",
                        "table"="test_tbl_varchar"
                    );

create external table jdbc_ext_table_mysql_int (
                    type_int tinyint
                    ) ENGINE=jdbc PROPERTIES (
                        "resource"="jdbc_ext_tbl_resource_outofrange",
                        "table"="test_tbl_int"
                    );
3. SELECT  '....'
select * from jdbc_ext_table_mysql_char;
select * from jdbc_ext_table_mysql_varchar;
select * from jdbc_ext_table_mysql_int;

### Expected behavior (Required)
report errors
### Real behavior (Required)
Empty
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
+----------------------+
| current_version()    |
+----------------------+
| MAIN-RELEASE 9da8e3f |
+----------------------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #6030

### CODE REVIEW

**Comment by silverbullet233** [comment]
@Mergifyio backport branch-2.2

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 1]
```diff
@@ -69,6 +69,10 @@ bool ScanOperator::has_output() const {
     if (_is_finished) {
         return false;
     }
+    // if storage layer returns an error, we should make sure `pull_chunk` has a chance to get it
+    if (Status status = get_scan_status(); !status.ok()) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/scan_operator.cpp:73]:
```suggestion
    if (!get_scan_status().ok()) {
```

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 2]
```diff
@@ -107,6 +111,10 @@ bool ScanOperator::is_finished() const {
     if (_is_finished) {
         return true;
     }
+    // if storage layer returns an error, we should make sure `pull_chunk` has a chance to get it
+    if (Status status = get_scan_status(); !status.ok()) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/scan_operator.cpp:115]:
```suggestion
    if (!get_scan_status().ok()) {
```

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/pipeline/scan/scan_operator.h [hunk 3]
```diff
@@ -69,6 +82,8 @@ class ScanOperator : public SourceOperator {
     std::atomic<int> _num_running_io_tasks = 0;
     std::vector<std::atomic<bool>> _is_io_task_running;
     std::vector<ChunkSourcePtr> _chunk_sources;
+    mutable SpinLock _scan_status_mutex;
+    Status _scan_status;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/scan_operator.h:86]: why need a special scan status, if storage engine return error status, we could pass the status to pipeline driver.

Comment (silverbullet233) [reply be/src/exec/pipeline/scan/scan_operator.h:86]: > why need a special scan status, if storage engine return error status, we could pass the status to pipeline driver.

this error is returned by chunk source in io threads which can not be get directly by pipeline driver, we should record it first

**Review by kangkaisen — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by silverbullet233** [comment]
@Mergifyio backport branch-2.1

**Pull request merged**